### PR TITLE
[image_picker] Check camera permissions and return error

### PR DESCRIPTION
**Pull request by wwwdata**
## Description

Under iOS, the camera permissions have not been correctly checked. If the user did not allow camera access, the camera view was launched and displayed a black screen. It was then also possible to make a photo of the black screen and use it. The camera view controller is now only launched, when the user has granted permissions. Otherwise it will not launch and flutter will also receive an error.

## Related Issues

https://github.com/flutter/flutter/issues/30378

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/image_picker/ios/Classes/ImagePickerPlugin.m [hunk 1]
```diff
@@ -118,6 +119,42 @@ - (void)showCamera {
...
+            }
+          });
+        } else {
+          dispatch_async(dispatch_get_main_queue(), ^{
+            [self errorNoAccess];
+          });
+        }
+      }];
+    }; break;
+    case AVAuthorizationStatusDenied:
```
Comment (cyanglaz) [inline packages/image_picker/ios/Classes/ImagePickerPlugin.m]: It is prob a good idea to send the failed statuses to the dart code so the developers can have more details on what caused the failure.

File: packages/image_picker/ios/Classes/ImagePickerPlugin.m [hunk 2]
```diff
@@ -118,6 +119,42 @@ - (void)showCamera {
...
+    case AVAuthorizationStatusRestricted:
+    default:
+      [self errorNoAccess];
+      break;
+  }
+}
+
+- (void)errorNoAccess {
+  _result([FlutterError errorWithCode:@"camera_access_denied"
```
Comment (cyanglaz) [inline packages/image_picker/ios/Classes/ImagePickerPlugin.m]: maybe the error code can be more specific on which case caused the failure?

**Comment by wwwdata** [comment]
Thank you for the super fast response. Makes a lot of sense I added your requested changes so flutter now knows if the user denied access or if the access was restricted altogether. Also formatted the code with clang-format so hopefully the build will be green.

File: packages/image_picker/CHANGELOG.md [hunk 3]
```diff
@@ -153,6 +158,7 @@
 
 * Updated dependencies
 * **Breaking Change**: You need to add a maven section with the "https://maven.google.com" endpoint to the repository section of your `android/build.gradle`. For example:
+
```
Comment (cyanglaz) [inline packages/image_picker/CHANGELOG.md]: Please remove this line.

Comment (wwwdata) [reply packages/image_picker/CHANGELOG.md]: oops that was my prettier plugin. Rebased, fixed and pushed again.

**Comment by cyanglaz** [comment]
@wwwdata 
Thank you for the update. Sorry about the back and force. There is one more thing that I would like to mention. Since adding this change makes `showCamera` being async, we need to add a guard to prevent `[_viewController presentViewController:_imagePickerController animated:YES completion:nil];` in `showCamera` being called twice. You can prob get a nasty result by doing this: Fresh install the app, hit the camera button twice very fast. Now both `checkAuthorization` gets triggered and it goes into `AVAuthorizationStatusNotDetermined`. And you wait for a second and hit grant access. 
A simple fix for this could be something like 
```
- (void)showCamera {
  @synchronized (self) {
    if (_imagePickerController.beingPresented) {
      return;
    }
    // Camera is not available on simulators
    if ([UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera]) {
      _imagePickerController.sourceType = UIImagePickerControllerSourceTypeCamera;
      [_viewController presentViewController:_imagePickerController animated:YES completion:nil];
    } else {
      [[[UIAlertView alloc] initWithTitle:@"Error"
                                  message:@"Camera not available."
                                 delegate:nil
                        cancelButtonTitle:@"OK"
                        otherButtonTitles:nil] show];
    }
  }
}
```
I know it's a bit more to ask. If you don't have to time to test and work on this, I am happy to shepherd this PR from here.

**Comment by wwwdata** [comment]
@cyanglaz very good catch! This was also a bug that was already present in the plugin. Even without calling it asynchronously (by granting camera permission, so it would directly call the code to launch the view) I could make it crash by rapidly tapping a button in flutter to launch the camera view. Tested it, and it doesn't crash anymore now.

**Comment by cyanglaz** [comment]
@wwwdata Thank you so much! Sorry that I would like to change one more thing which is the version. This change actually changing the behavior a little to the public API. (Now we can catch the error when trying to launch the image picker on iOS). So we have to increment the the version to 0.5.2.  You might also want to mention that in the documentation of `pickImage`, but I can do that documentation after we merge in this PR if you don't feel like doing it. 
As long as you can update the version to 0.5.2, I will merge this PR!
Again, thanks for fixing this!

**Comment by wwwdata** [comment]
OK. I am not so sure how to exactly document this so I will leave it up to you. Thank you too for the fast and good feedback. My next pull request is hopefully a lot smoother then :)

**Pull request merged**